lib/vm-agent.js

```diff
@@ -9,450 +9,696 @@
+ * delay and then that VM's uuid is added back to the queue (only if the uuid
+ * is not already queued) and we will re-run this process when it is next loaded
+ * from the queue. The delay before re-queueing will double on each failure, up
+ * to some maximum value (MAX_UPDATE_DELAY) but the delay will be reset on a
+ * sucessful update.
+ *
+ *
+ * Important Notes
+ * ===============
+ *
+ * VMs which have the 'do_not_inventory' property set are not visible to
+ * vm-agent. They should be treated exactly the same as VMs which do not exist
+ * on this CN.
+ *
  */
 
-var async = require('async');
-var backoff = require('backoff');
+var fs = require('fs');
+var path = require('path');
+
+var assert = require('assert-plus');
+var diff = require('deep-diff').diff;
+var vasync = require('vasync');
+var vmadm = require('vmadm');
 
-var VM = require('./vm');
+var VmWatcher = require('./vm-watcher');
 var VMAPI = require('./vmapi-client');
 
-var _versionCache = null;
-function getAgentVersion() {
-    if (_versionCache === null) {
-        _versionCache = require('../package.json').version;
-    }
-    return _versionCache;
-}
+// initial and maximum values to delay between VMAPI retries. (in ms)
+var INITIAL_UPDATE_DELAY = 500;
+var MAX_UPDATE_DELAY = 30000;
+
 
 function VmAgent(options) {
-    this.options = options;
-    this.log = options.log;
-    this.sample = null;
-    this.lastFullSample = null;
-    this.uuid = options.uuid;
-    this.version = getAgentVersion();
-
-    var userAgent = 'vm-agent/' + this.version +
-        ' (' + 'node/' + process.versions.node + ')' +
-        ' server/' + this.uuid;
-
-    this.vmapiClient = new VMAPI({
+    var self = this;
+    var packageJson = path.dirname(__dirname) + '/package.json';
+    var userAgent;
+
+    assert.object(options, 'options');
+    assert.object(options.log, 'options.log');
+    assert.optionalNumber(options.periodicInterval, 'options.periodicInterval');
+    assert.uuid(options.server_uuid, 'options.server_uuid');
+    assert.string(options.url, 'options.url');
+
+    // Load the list of fields that VmWatcher will be watching, which we'll
+    // use when comparing objects. Since VMAPI and vmadm have different default
+    // fields, this is the common list. We start with just those fields that
+    // we're actually watching for changes.
+    self.comparisonFields = VmWatcher.WATCHED_FIELDS;
+
+    // We add boot_timestamp to notice case where zone has rebooted between
+    // events.
+    if (self.comparisonFields.indexOf('boot_timestamp') === -1) {
+        self.comparisonFields.push('boot_timestamp');
+    }
+
+    if (options.periodicInterval) {
+        self.periodicInterval = options.periodicInterval;
+    }
+
+    self.log = options.log;
+    self.server_uuid = options.server_uuid;
+    self.version = JSON.parse(fs.readFileSync(packageJson)).version;
+
+    assert(self.version, 'missing package.json version');
+
+    userAgent = 'vm-agent/' + self.version
```

Comment:
@misterdjules not at this point, that I'm aware of. In the future we did talk about restricting access to `VMAPI PUT /vms` such that it could only be used by vm-agent and the jobs that are using cn-agent and loading the latest state on completion. No ticket for that yet though.